RapidiaTechInc/Marlin
# Stop and reset

## Details

This PR has the following effects:

- sending byte `\x04` ("End Of Transmission") at any point over serial is the same as sending M112.
- `R750` is introduced, which resets Marlin immediately.
- emergency parser remains active even after Marlin is killed (e.g. by M112), allowing `R750` to be used to reset the device.

## Implementation

R750 is implemented by jumping to 0x0, which is what the reset button and watchdog timer do.

Emergency parser remains active by changing the way that kill() disables interrupts. Normally, kill simply clears the interrupt flag (`cli()`) and then loops forever resetting the watchdog timer. In order to keep the serial communication channel open, `cli()` cannot be used. In its place, all interrupts are individually disabled and their pending interrupt flags cleared. (see `kill_recovery.cpp`), with the exception of the serial communication interrupts and the `millis()` interrupt.

## Testing Performed

- Unit test verifies that no interrupts are defined anywhere in the firmware which are not listed in `kill_recovery.cpp`
- Isolated arduino without peripherals works as expected.

## Testing Required
- Try running M112 on an actual metal printer during a print and ensure that all heaters, lamp, and steppers turn off and that power is shut down. 
- Check that the R750 then is able to reset the printer to a working state.
- Please note that M112 and R750 do not currently respond with `ok`.
